ppa/archive/management/commands/hathi_images.py

```diff
@@ -0,0 +1,165 @@
+from django.template.defaultfilters import pluralize
+from corppa.utils.path_utils import encode_htid, get_vol_dir
+
+from ppa.archive.models import DigitizedWork
+from ppa.archive.templatetags.ppa_tags import page_image_url
+
+
+class Command(BaseCommand):
+    """
+    Download HathiTrust page image data via image server
+
+    Note: Excerpts cannot be specified individually, only by source (collectively)
+    """
+    help = __doc__
+    #: normal verbosity level
+    v_normal = 1
+    verbosity = v_normal
+    #: crawl delay (in seconds)
+    crawl_delay=1
+    def add_arguments(self, parser):
+        parser.add_argument(
+            "out",
+            type=Path,
+            help="Top-level output directory")
+        parser.add_argument(
+            "--htids",
+            nargs="*",
+            help="Optional list HathiTrust ids to download",
+        )
+        parser.add_argument(
+            "--progress",
+            action="store_true",
+            help="Display progress bars to track download progress",
+            default=True,
+        )
+    
+    def download_image(self, page_url: str, out_file: Path) -> None:
+        response = requests.get(page_url)
+        if response.status_code == requests.codes.ok:
+            with out_file.open(mode="wb") as writer:
+                writer.write(response.content)
+        # Apply crawl delay after request
+        sleep(self.crawl_delay)
+
+
+    def handle(self, *args, **kwargs):
+        self.verbosity = kwargs.get("verbosity", self.v_normal)
+        self.options = kwargs
+
+        # validate output directory
+        if not kwargs["out"]:
+            raise CommandError("An output directory must be specified")
+        output_dir = kwargs["out"]
+        if not output_dir.is_dir():
+            raise CommandError(
+                f"Output directory '{output_dir}' does not exist or is not a directory"
+            )
+
+        # use ids specified via command line when present
+        htids = kwargs.get("htids", [])
+
+        # by default, sync data for all non-suppressed hathi source ids
+        digworks = DigitizedWork.objects.filter(
+            status=DigitizedWork.PUBLIC, source=DigitizedWork.HATHI
+        )
+
+        # if htids are specified via parameter, use them to filter
+        # the queryset, to ensure we only sync records that are
+        # in the database and not suppressed
+        if htids:
+            digworks = digworks.filter(source_id__in=htids)
+
+        # bail out if there's nothing to do
+        # (e.g., explicit htids only and none valid)
+        if not digworks.exists():
+            self.stdout.write("No records to download; stopping")
+            return
+
+        # setup main progress bar
+        overall_progress = None
+        if self.options["progress"]:
+            overall_progress = progressbar.ProgressBar(
+                redirect_stdout=True, max_value=digworks.count(), max_error=False
```

Comment:
Ah, good call. It was hard finding example code, since most examples were for independent / multithreaded situations. But I'll look into it. Maybe we can find some time to make some design choices next week?